File: Medium/138. Copy List with Random Pointer/solution.java
class Solution {
    public Node copyRandomList(Node head) {
        Node ans = head;
        Node cur = head;
        while (cur != null) {
            Node next = cur.next;
            Node newNode = new Node(cur.val);
            cur.next = newNode;
            newNode.next = next;
            cur = next;
        }
        cur = head;
        while (cur != null) {
            if (cur.random != null) {
                cur.next.random = cur.random.next;
            }
            cur = cur.next.next;
        }
        Node dummy = new Node(0);
        Node dummyCopy = dummy;
        Node copy;
        cur = head;
        while (cur != null) {
            Node next = cur.next.next;
            copy = cur.next;
            dummyCopy.next = copy;
            dummyCopy = copy;
            cur.next = next;
            cur = next;
        }
        return dummy.next;
    }
}

File: Medium/138. Copy List with Random Pointer/solution.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':

        if not head:
            return None

        curr = head

        while curr:
            nxt = curr.next
            copy = Node(curr.val, None, None)
            curr.next = copy
            curr.next.next = nxt
            curr = curr.next.next
        
        curr = head

        while curr:
            if curr.random:
                curr.next.random = curr.random.next
            curr = curr.next.next
        

        ptr_old_list = head
        ptr_new_list = head.next
        new_head = head.next

        while ptr_old_list:
            ptr_old_list.next = ptr_old_list.next.next
            ptr_new_list.next = ptr_new_list.next.next if ptr_new_list.next else None

            ptr_old_list = ptr_old_list.next
            ptr_new_list = ptr_new_list.next
        
        return new_head

